Added real32 (Beckhoff float) and real64 (Beckhoff double)

JensVanhooydonck opened the pull request:
I need to set some values which have type REAL32.
https://infosys.beckhoff.com/english.php?content=../content/1033/ioanalogmanual/12855866763.html&id=

JensVanhooydonck committed:
Added real32 (Beckhoff float) https://infosys.beckhoff.com/english.php?content=../content/1033/ioanalogmanual/12855866763.html&id=

JensVanhooydonck commented:
Still need to test this. As soon as next monday.

yguel commented:
Hi JensVanhooydonck, 
Again, many thanks for all your PR.
All the checks did not pass, but the code sounds good and I agree with the need of the feature. 
As soon as the checks pass, I will merge the PR.

JensVanhooydonck committed:
Changed EC_READ_REAL and EC_WRITE_REAL to bitcasting. EC_READ_REAL and EC_WRITE_REAL aren't always available.

JensVanhooydonck commented:
It seems that the EC_READ_REAL and EC_WRITE_REAL funtions aren't always available. Used the same logic (bit casting) and reading and writing with the EC_READ_UINT32 and EC_WRITE_UINT32 functions.

For the sdo, there was currently only int value data available. I added a double data value as well. Should I keep it this way, or should I change the data type from int to double and cast the value to int for the existing sdo's?

Also, should I also add the double/real64 values as well?

JensVanhooydonck committed:
Working SDO + altered casting of floats + added doubles

yguel commented:
Hi JensVanhooydonck,
> It seems that the EC_READ_REAL and EC_WRITE_REAL functions aren't always available. Used the same logic (bit casting) and reading and writing with the EC_READ_UINT32 and EC_WRITE_UINT32 functions.

That is the way it should be done to work everywhere.

> 
> For the sdo, there was currently only int value data available. I added a double data value as well. Should I keep it this way, or should I change the data type from int to double and cast the value to int for the existing sdo's?


I think having different types is the ideal way to go, especially if we imagine having 64 bits integers at some point.

> 
> Also, should I also add the double/real64 values as well?


2 cases:
-  if there is no real use case, I prefer to wait for one and document how it could be done somewhere, to not work for nothing. But
-  if it is very very easy, like it takes you less than 5min, then it is the best way to document it, so let's do it.

JensVanhooydonck commented:
> Hi JensVanhooydonck,
> 
> > It seems that the EC_READ_REAL and EC_WRITE_REAL functions aren't always available. Used the same logic (bit casting) and reading and writing with the EC_READ_UINT32 and EC_WRITE_UINT32 functions.
> 
> That is the way it should be done to work everywhere.
> 
> > For the sdo, there was currently only int value data available. I added a double data value as well. Should I keep it this way, or should I change the data type from int to double and cast the value to int for the existing sdo's?
> 
> I think having different types is the ideal way to go, especially if we imagine having 64 bits integers at some point.
> 
> > Also, should I also add the double/real64 values as well?
> 
> 2 cases:
> 
> * if there is no real use case, I prefer to wait for one and document how it could be done somewhere, to not work for nothing. But
> * if it is very very easy, like it takes you less than 5min, then it is the best way to document it, so let's do it.

1. I've added all logic (casting to U32 and U64) because the build in functions didn't work for me. (Gave a runtime error, building was ok). Didn't figure out why it didn't find the functions.

2. Currently i've altered the code so when setting the sdo values, this is already casted to the int data value. If we need it in the future to also support 64-bit ints we need to fix it then i assume.

3. For double/real64 in PDO i've added the code already.


I've tested this with real32 already and everything seems to work correctly. Don't have a device with Real64 items currently. If needed, i can comment this part out.

JensVanhooydonck committed:
Removed unused code part SDO

christophfroehlich commented on ethercat_interface/include/ethercat_interface/ec_pdo_channel_manager.hpp:
```diff
@@ -92,6 +100,13 @@ class EcPdoChannelManager
+      uint32_t raw = *(uint32_t *)&f;
+      EC_WRITE_U32(domain_address, static_cast<uint32_t>(raw));
+    } else if (data_type == "real64" || data_type == "double") {
+      uint32_t raw = *(uint32_t *)&value;
+      EC_WRITE_U64(domain_address, static_cast<uint64_t>(raw));
```
```suggestion
      uint64_t raw = *(uint64_t *)&value;
      EC_WRITE_U64(domain_address, static_cast<uint64_t>(raw));
```

christophfroehlich commented on ethercat_interface/include/ethercat_interface/ec_sdo_manager.hpp:
```diff
@@ -79,7 +79,15 @@ class SdoConfigEntry
+        float floatvalue = sdo_config["value"].as<float>();
+        data = *(int *)&floatvalue;
+      } else if (data_type == "double" || data_type == "real64") {
+        float doublevalue = sdo_config["value"].as<double>();
```
```suggestion
        double doublevalue = sdo_config["value"].as<double>();
```

christophfroehlich reviewed (commented):
I think that there are some copy-paste errors for double type?

tpoignonec commented:
Hi,
FYI, I merged #197 and will backport it to the `humble` branch.
I'll close this PR and directly cherry pick from the other PR.
